addons/mass_mailing_sms/models/mail_notification.py

```diff
@@ -8,51 +8,43 @@
 from odoo.osv import expression
 
 
-class MailingTrace(models.Model):
+class MailNotification(models.Model):
     """ Improve statistics model to add SMS support. Main attributes of
     statistics model are used, only some specific data is required. """
-    _inherit = 'mailing.trace'
+    _inherit = 'mail.notification'
     CODE_SIZE = 3
 
-    trace_type = fields.Selection(selection_add=[('sms', 'SMS')])
-    sms_sms_id = fields.Many2one('sms.sms', string='SMS', index=True, ondelete='set null')
-    sms_sms_id_int = fields.Integer(
+    sms_id_int = fields.Integer(
         string='SMS ID (tech)',
         help='ID of the related sms.sms. This field is an integer field because '
              'the related sms.sms can be deleted separately from its statistics. '
              'However the ID is needed for several action and controllers.',
         index=True,
     )
-    sms_number = fields.Char('Number')
     sms_code = fields.Char('Code')
     failure_type = fields.Selection(selection_add=[
-        ('sms_number_missing', 'Missing Number'),
-        ('sms_number_format', 'Wrong Number Format'),
-        ('sms_credit', 'Insufficient Credit'),
-        ('sms_server', 'Server Error'),
-        # mass mode specific codes
         ('sms_blacklist', 'Blacklisted'),
         ('sms_duplicate', 'Duplicate'),
     ])
 
     @api.model_create_multi
     def create(self, values_list):
         for values in values_list:
-            if 'sms_sms_id' in values:
-                values['sms_sms_id_int'] = values['sms_sms_id']
-            if values.get('trace_type') == 'sms' and not values.get('sms_code'):
+            if 'sms_id' in values:
+                values['sms_id_int'] = values['sms_id']
+            if values.get('notification_type') == 'sms' and not values.get('sms_code'):
                 values['sms_code'] = self._get_random_code()
-        return super(MailingTrace, self).create(values_list)
+        return super(MailNotification, self).create(values_list)
 
     def _get_random_code(self):
         """ Generate a random code for trace. Uniqueness is not really necessary
         as it serves as obfuscation when unsubscribing. A valid trio
         code / mailing_id / number will be requested. """
         return ''.join(random.choice(string.ascii_letters + string.digits) for dummy in range(self.CODE_SIZE))
 
-    def _get_records_from_sms(self, sms_sms_ids=None, additional_domain=None):
-        if not self.ids and sms_sms_ids:
-            domain = [('sms_sms_id_int', 'in', sms_sms_ids)]
+    def _get_records_from_sms(self, sms_ids=None, additional_domain=None):
+        if not self.ids and sms_ids:
+            domain = [('sms_id_int', 'in', sms_ids)]
         else:
             domain = [('id', 'in', self.ids)]
```

Comment:
Wonder what will happen if you have an empty recordset.

Maybe check for `self.ids` and return `self.env['mail.notification']` if it's empty?

Same in `_get_records`

Reply:
The search method doesn't make the trick ?